Implement Sphinx documentation with GitHub Pages auto-deployment (#8)
nasa/pysips: docs/tutorial.rst

   mse = mean_squared_error(y_test, y_pred)
   
   print(f"Discovered expression: {expression}")
   print(f"R² score: {r2:.4f}")
   print(f"MSE: {mse:.4f}")

.. _custom-operators:

3. Customizing Available Operators
-----------------------------------

You can control which mathematical operators are available:

.. code-block:: python

   from pysips import PysipsRegressor

   # Example 1: Simple arithmetic only
   regressor_simple = PysipsRegressor(
       operators=['+', '-', '*'],  # No division
       max_complexity=10
   )

   # Example 2: Include division and power
   regressor_extended = PysipsRegressor(
       operators=['+', '-', '*', '/', 'pow'],
       max_complexity=15
   )

   # Example 3: Include trigonometric functions
   regressor_trig = PysipsRegressor(
       operators=['+', '-', '*', 'sin', 'cos'],
       max_complexity=15
   )

   # Example 4: Include exponential and logarithm
   regressor_exp = PysipsRegressor(
       operators=['+', '-', '*', 'exp', 'log'],
       max_complexity=15
   )

**Available operators:**

- Arithmetic: ``'+'``, ``'-'``, ``'*'``, ``'/'``
- Power: ``'pow'``, ``'sqrt'``
- Trigonometric: ``'sin'``, ``'cos'``, ``'tan'``

> - Exponential/Logarithmic: ``'exp'``, ``'log'``, ``'log10'``

Reviewer: ```suggestion
- Exponential/Logarithmic: ``'exp'``, ``'log'``
```

       operators=['+', '-', '*', 'pow'],
       max_complexity=15,
       num_particles=200,
       num_mcmc_samples=10,
       max_time=120,
       random_state=42
   )
   regressor.fit(X, y)

   # Get all sampled models and their likelihoods
   models, likelihoods = regressor.get_models()

   print(f"Total unique models sampled: {len(models)}")
   print(f"\nTop 5 models by likelihood:")
   
   # Sort by likelihood
   sorted_indices = np.argsort(likelihoods)[::-1]
   
   for i in range(min(5, len(models))):
       idx = sorted_indices[i]
       print(f"{i+1}. {models[idx]} (likelihood: {likelihoods[idx]:.4f})")

   # Count frequency of each model in the posterior
   print(f"\nSelected model: {regressor.get_expression()}")
   print(f"(This is the most frequently sampled model)")

Advanced Topics
---------------

**Mutation and Crossover Control:**

Fine-tune the proposal mechanism probabilities:

.. code-block:: python

   regressor = PysipsRegressor(
       operators=['+', '-', '*'],
       # Mutation probabilities
       command_probability=0.3,    # Probability of changing an operator
       node_probability=0.3,        # Probability of replacing a node

>        parameter_probability=0.2,   # Probability of modifying a constant

Reviewer: ```suggestion
       parameter_probability=0.2,   # Probability of changing edges in an expression graph
```

   )

Best Practices
--------------

1. **Start simple**: Begin with basic operators and small complexity limits
2. **Use checkpoints**: For experiments taking >5 minutes, always use checkpointing
3. **Reproducibility**: Always set ``random_state`` for reproducible results
4. **Progress monitoring**: Use ``show_progress_bar=True`` (default) for interactive use
5. **Hyperparameter tuning**: Disable progress bar when using GridSearchCV
6. **Model validation**: Always validate on held-out test data
7. **Posterior inspection**: Check the full posterior distribution, not just the best model

Troubleshooting
---------------

**Issue: Fitting is too slow**

- Reduce ``num_particles`` (e.g., 50-100)
- Reduce ``max_complexity`` (e.g., 10-15)
- Set ``max_time`` to limit runtime
- Simplify the operator set

**Issue: Poor model quality**

- Increase ``num_particles`` (e.g., 200-500)
- Increase ``num_mcmc_samples`` (e.g., 20-50)
- Increase ``max_complexity`` if needed
- Add more relevant operators
- Collect more/better training data

**Issue: Overfitting**

- Use cross-validation for model selection
- Reduce ``max_complexity``
- Use ``model_selection='mode'`` instead of ``'max_likelihood'``

> - Add regularization by penalizing complex expressions

Reviewer: ```suggestion
```

**Issue: Checkpoint file corrupted**

- Delete the checkpoint file and start fresh
- Ensure sufficient disk space
- Check file permissions

> 

Reviewer: ```suggestion
```

Further Resources
-----------------

- `GitHub Repository <https://github.com/nasa/pysips>`_
- `API Reference <api/modules.html>`_

> - `Paper: Bayesian Symbolic Regression via Posterior Sampling <https://royalsocietypublishing.org/>`_

Reviewer: ```suggestion
- `Paper: Bayesian Symbolic Regression via Posterior Sampling <https://arxiv.org/abs/2512.10849>`_
```

> - `Paper: Bayesian Symbolic Regression via Posterior Sampling <https://royalsocietypublishing.org/>`_

Reviewer: lets use the arxiv link for now
